muxado/src/session.rs

```diff
@@ -143,12 +145,34 @@ where
             open_reqs: open_rx,
         };
 
-        tokio::spawn(read_task.run());
-        tokio::spawn(write_task.run());
+        let (dropref, waiter) = awaitdrop::awaitdrop();
+
+        tokio::spawn(
+            futures::future::select(
+                async move {
+                    let result = read_task.run().await;
+                    debug!(?result, "read_task exited");
+                }
+                .boxed(),
+                waiter.wait(),
+            )
+            .instrument(debug_span!("read_task")),
+        );
+        tokio::spawn(
+            futures::future::select(
+                async move {
+                    let result = write_task.run().await;
+                    debug!(?result, "write_task exited");
+                }
+                .boxed(),
+                waiter.wait(),
+            )
+            .instrument(debug_span!("write_task")),
+        );
```

Comment:
A tiny bit surprised you resisted using a macro :)

Reply:
My threshold is 3 :wink: 

Or at least greater than 2 lol